bot/exts/moderation/clean.py

```diff
@@ -0,0 +1,589 @@
+
+        if first_limit and channels and (channels == "*" or len(channels) > 1):
+            raise BadArgument("Message or time range specified across multiple channels.")
+
+        if (isinstance(first_limit, Message) or isinstance(second_limit, Message)) and channels:
+            raise BadArgument("Both a message limit and channels specified.")
+
+        if isinstance(first_limit, Message) and isinstance(second_limit, Message):
+            # Messages are not in same channel.
+            if first_limit.channel != second_limit.channel:
+                raise BadArgument("Message limits are in different channels.")
+
+        if users and bots_only:
+            raise BadArgument("Marked as bots only, but users were specified.")
+
+        # This is an implementation error rather than user error.
+        if second_limit and not first_limit:
+            raise ValueError("Second limit specified without the first.")
+
+    @staticmethod
+    async def _send_expiring_message(ctx: Context, content: str) -> None:
+        """Send `content` to the context channel. Automatically delete if it's not a mod channel."""
+        delete_after = None if is_mod_channel(ctx.channel) else MESSAGE_DELETE_DELAY
+        await ctx.send(content, delete_after=delete_after)
+
+    @staticmethod
+    def _build_predicate(
+        bots_only: bool = False,
+        users: list[User] = None,
+        regex: Optional[str] = None,
+        first_limit: Optional[datetime] = None,
+        second_limit: Optional[datetime] = None,
+    ) -> Predicate:
+        """Return the predicate that decides whether to delete a given message."""
+        def predicate_bots_only(message: Message) -> bool:
+            """Return True if the message was sent by a bot."""
+            return message.author.bot
+
+        def predicate_specific_users(message: Message) -> bool:
+            """Return True if the message was sent by the user provided in the _clean_messages call."""
+            return message.author in users
+
+        def predicate_regex(message: Message) -> bool:
+            """Check if the regex provided in _clean_messages matches the message content or any embed attributes."""
+            content = [message.content]
+
+            # Add the content for all embed attributes
+            for embed in message.embeds:
+                content.append(embed.title)
+                content.append(embed.description)
+                content.append(embed.footer.text)
+                content.append(embed.author.name)
+                for field in embed.fields:
+                    content.append(field.name)
+                    content.append(field.value)
+
+            # Get rid of empty attributes and turn it into a string
+            content = "\n".join(attr for attr in content if attr)
+
+            # Now let's see if there's a regex match
+            return bool(re.search(regex, content, re.IGNORECASE + re.DOTALL))
```

Comment:
We should catch for `re.error` here in case the regex is invalid.

---

Example: !clean regex `[`

```py
Traceback (most recent call last):
  File "/home/shivansh/.cache/pypoetry/virtualenvs/bot-5X3Lakqr-py3.9/lib/python3.9/site-packages/discord/ext/commands/core.py", line 85, in wrapped
    ret = await coro(*args, **kwargs)
  File "/home/shivansh/Programming/pydis_projects/bot/bot/exts/moderation/clean.py", line 521, in clean_regex
    await self._clean_messages(ctx, traverse, regex=regex, channels=channels, use_cache=use_cache)
  File "/home/shivansh/Programming/pydis_projects/bot/bot/exts/moderation/clean.py", line 394, in _clean_messages
    message_mappings, message_ids = await self._get_messages_from_channels(
  File "/home/shivansh/Programming/pydis_projects/bot/bot/exts/moderation/clean.py", line 227, in _get_messages_from_channels
    if to_delete(message):
  File "/home/shivansh/Programming/pydis_projects/bot/bot/exts/moderation/clean.py", line 154, in predicate_regex
    return bool(re.search(regex, content, re.IGNORECASE + re.DOTALL))
  File "/home/shivansh/.pyenv/bin/versions/3.9.5/lib/python3.9/re.py", line 201, in search
    return _compile(pattern, flags).search(string)
  File "/home/shivansh/.pyenv/bin/versions/3.9.5/lib/python3.9/re.py", line 304, in _compile
    p = sre_compile.compile(pattern, flags)
  File "/home/shivansh/.pyenv/bin/versions/3.9.5/lib/python3.9/sre_compile.py", line 764, in compile
    p = sre_parse.parse(p, flags)
  File "/home/shivansh/.pyenv/bin/versions/3.9.5/lib/python3.9/sre_parse.py", line 948, in parse
    p = _parse_sub(source, state, flags & SRE_FLAG_VERBOSE, 0)
  File "/home/shivansh/.pyenv/bin/versions/3.9.5/lib/python3.9/sre_parse.py", line 443, in _parse_sub
    itemsappend(_parse(source, state, verbose, nested + 1,
  File "/home/shivansh/.pyenv/bin/versions/3.9.5/lib/python3.9/sre_parse.py", line 549, in _parse
    raise source.error("unterminated character set",
re.error: unterminated character set at position 0
```

Reply:
Right, I was going to do that but forgot, thanks.